### ISSUE TITLE
Map protection: mixed CR and LF against editing

### ISSUE THREAD

**Issue by Luashine** [lep/pjass#11]
> I found why you couldn't modify the map before the map has used MIXED line endings, some CR lines, some are LF lines. Because rawcodes sometimes contained LF, you can't just convert the whole file's line endings. But if there's a CR line too long it will crash the game/script. Notepad++ detected the file as CR.
I think you can circumvent the detection (LF is preferred, but CRLF works too) if you replace the CR line endings in globals block at the beginning with LF -- notepad++: search and replace & replace in selection only & extended: search `\r` replace: `\n`

---

Now for the pjass: #6 - there I tested converting the whole file to` \r` and it just crashed the game.

In current case the map protection interchangeably used `\r` and `\n` for lines. On one hand `\r` apparently counts as a line separator for the interpreter, so you can have a two statements with `line1\rline2\n`. At the same time, the protection used `\n` too and thus avoided crashing the game like in #6 (too long `\r` line)

I wanted to document it in some capacity first and foremost. Notepad++ is troublesome if it detects the file as CR because then you can't paste any long code into the map because it will produce a "long CR line" (and in the end cheaters are still not stopped by that...)

pjass does not report any errors, technically it's correct. The game is alright with that. But I think there's a **different limit** similar to #6 where pjass didn't report me any errors yet but the game parsed the Jass wrongly.

> 6/22 12:23:21.585  Opening map - C:/Users/shiny/Documents/Warcraft III/Maps/Jassdoc/UpgradeRPG-UnitTest/Upgrade_RPG_7.0G_Reporged Eng-inst.w3x
6/22 12:23:22.326  Map contains invalid Jass scripts that couldn't be compiled by war3, file: war3map.j @ 3378, error: encountered undeclared identifier 'r' on line 3378

This error happened right after I added the following code right after "endglobals" in CR mode using Notepad++:

```jass
function CreateUnitLogging takes player p, integer rawcode, real x, real y, real face returns unit
    local unit u
    local integer createdId
    local string unitModel
	local string funcCallString
    
    set u = CreateUnit(p,rawcode,x,y,face)
    set createdId = GetUnitTypeId(u)
	set funcCallString = "call CreateUnit("+ I2S(GetPlayerId(p)) +","+ I2S(rawcode) +","+R2S(x)+","+R2S(y)+","+R2S(face)+")"
    
    // patch 1.31.0.11889
    set unitModel = BlzGetUnitStringField(u, ConvertUnitStringField ('umdl'))
    
    call BJDebugMsg(funcCallString + " & model: '" + unitModel + "'")
    
    if rawcode != createdId then
		call BJDebugMsg("MISMATCH!!! last unit has id : " + I2S(createdId))
    endif
    
    // this leaks because we cannot clear and return u at the same time without a global or another function
    return u
endfunction
```

I will attach the map's blizzard.j and war3map.j (MPQ:scripts/ folder) as compressed zip. Looks like the protection is easy to encounter, at least two versions of Upgrade RPG (Korean) have it. The blizzard.j file is based on 1.24+ version, it has some additions. The Reporged version doesn't need Dz API or anything of that kind and runs on 1.36.2

> Upgrade_RPG_7.0G_Reporged Eng.w3x

[Scripts.zip](https://github.com/user-attachments/files/15936266/Scripts.zip)

---

What to do:

1. Find out the exact limit where this begins to cause problems (after pasting the above code in CR mode)
2. If possible make pjass warn about mixed CR & LF in a file. Not CRLF but separately used CR and LF. A general once per file warning or if (1) is found an exact **error** message too. Ah I remember there was a problem with warnings vs errors right? :)

**Comment by lep** [comment]
Hmm, i assume you're talking about the most recent patch, that is you tested that X-Hero map on current BNet. As you said, it's as easy as reverting that commit so wen can just do that at any point but i would like to understand the issue. Because i think i tried using CRs and it didn't work.

**Comment by Slayer95** [comment]
> Hmm, i assume you're talking about the most recent patch, that is you tested that X-Hero map on current BNet

No. I tested it to work in 1.27 (note that #6 claimed that CR didn't work in both 1.27 and 1.33.) I'll try to check whether it works in 2.02 in Single Player.

As far as I can tell, the observations in #6 are the result of a combination of:
1. Using Git to store ``war3map.j`` files in text mode, when it should be binary.
2. Manipulation of JASS files using FourCC codes containing 0xA and/or 0xD.

Of course, changing a FourCC character from CR to LF. Adding a fifth char, or converting ``\x0d\x0a`` to ``\x0a`` are all operations that result in either invalid JASS, or invalid handles.

**Comment by Luashine** [comment]
Thanks, it's a good test map. The only places where it uses LF is inside multi-line strings.

The script is minified, but not obfuscated. It doesn't contain comments. What if CR only breaks with comments or when followed up by blank space / tab after CR?

**Comment by Slayer95** [comment]
(1.27b)

We live to run another line of code - proven.

![Image](https://github.com/user-attachments/assets/eec213c2-d09c-4096-b794-188c235e64e1)

I also ran tests to check whether CR or LF are required before ``//`` to start a comment, and it turns out the engine accepts either or both before ``//``, and it gets it's a comment.

![Image](https://github.com/user-attachments/assets/a5cdd2f5-9a0c-47a1-8b9f-41615fc4bd88)

I also checked inserting CR, LF or both inside a string, and got surprising results (either are legal, but join them and it crashes!)

![Image](https://github.com/user-attachments/assets/83603856-7ccf-4c6f-8a9b-2dad7be36060)

[test-cases.zip](https://github.com/user-attachments/files/20872935/test-cases.zip)

Notes:
- You may notice that I literally wrote ``(LF)`` in the JASS script. That was originally unintended, but the fact that WC3 doesn't complain means that the ``call`` line after ``CR`` is indeed treated as part of the comment.
- Lone ``2100).`` messages are because I carelessly put a ``CR`` after a ``//Zoom`` comment, therefore causing ``udg_RFTextCmdZoomNoticePrefix`` to not be changed from its default ``""`` value to what the map originally intended.

**Comment by Slayer95** [comment]
> Lone 2100). messages are because I carelessly put a CR after a //Zoom comment, therefore causing udg_RFTextCmdZoomNoticePrefix to not be changed from its default "" value to what the map originally intended.

I am re-running the tests for CR/LF/CRLF within a string, because I realized this is a critical issue.

**Comment by Slayer95** [comment]
Notably, after reverting 8e138d9e70a6d9ec7906e7df858da90e71d5c146,

- ``comment-CR-before-statement-ignored`` (should pass), fails at L8531 with "syntax error".
- ``comment-CR-before-block-crash`` (should fail) fails ~200 lines later with a random "too long string literal" error.

![Image](https://github.com/user-attachments/assets/7ce102b3-b420-488d-bc8b-2a1104cc02c0)

**Comment by lep** [comment]
Ok, so here is my understanding of the whole situation. And please correct me if I'm wrong: CR line-endings do in fact work **but** comments have to end in LF. So i would revert that mentionend commit and apply the following diff to pjass.

```diff
diff --git a/token.l b/token.l
index 9793478..57af2dc 100644
--- a/token.l
+++ b/token.l
@@ -53,11 +53,11 @@ ID       [a-zA-Z]([a-zA-Z0-9_]*[a-zA-Z0-9])?
     BEGIN(xCOMMENT);
 }
 
-<xCOMMENT>[^\r\n]* { }
+<xCOMMENT>[^\n]* { }
 <xCOMMENT>{NEWLINE} { BEGIN(INITIAL); return NEWLINE; }
 <xCOMMENT><<EOF>> { BEGIN(INITIAL); return NEWLINE; }
 
-<xANNOTATION>[^\r\n]* { annotations = updateannotation(annotations, yytext, &available_flags); }
+<xANNOTATION>[^\n]* { annotations = updateannotation(annotations, yytext, &available_flags); }
 <xANNOTATION>{NEWLINE} { BEGIN(INITIAL); return NEWLINE; }
 <xANNOTATION><<EOF>> { BEGIN(INITIAL); return NEWLINE; }
 

```

**Comment by Luashine** [comment]
Oversimplified, but correct. What we forgot to test is multiple CR comment lines followed by `//last comment\r\n`

Otherwise, being overly pedantic, a CR comment line is still legal under specific circumstances (which cannot be implemented statelessly imho)

**Comment by Luashine** [comment]
Do you want me to write down test cases now? I'm only on mobile, it'd take a while to type it out. Only 2/4 tests in that commit seem correct based on current understanding @lep

**Comment by lep** [comment]
I mean the tests were added with my understanding at that time so it's ok that they would fail once i change the code. I would review the tests once i made the changes of course. You can still provide (probably better) test cases if you want; i would be glad. But i don't think there is any hurry.

**Comment by Luashine** [comment]
@lep looks complete. 👍 
If you want to avoid the dreaded git EOL conversions: .gitattributes

`path/to/file -text` negate text attr, but it keeps the diff ability.

**Comment by Luashine** [comment]
Turns out an earlier "fix" from 2016 is what allowed the buggy combination of COMMENT + CR :) https://github.com/lep/pjass/commit/d822cacbd889fe723102102c8a09d578d90f013e#diff-527c39601153c6fcdc9d943a2aaea7bbc98b16b086306a5fb64587769d2e647eR1
